Ghidra/Processors/ARC/data/languages/ARCompact.slaspec

```diff
@@ -0,0 +1,1793 @@
+      op4_b_dst = op4_c_src + op4_c_src;
+    }
+    :asl.f op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x00 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      update_flags_add_cv(op4_c_src, op4_c_src);
+      op4_b_dst = op4_c_src + op4_c_src;
+      update_flags_zn(op4_b_dst);
+    }
+    :asr op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x01 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = op4_c_src s>> 1;
+    }
+    :asr.f op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x01 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      local lsb:4 = op4_c_src & 1:4;
+      C = lsb:1;
+      op4_b_dst = op4_c_src s>> 1;
+      update_flags_zn(op4_b_dst);
+    }
+    :lsr op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x02 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = op4_c_src >> 1;
+    }
+    :lsr.f op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x02 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      local lsb:4 = op4_c_src & 1:4;
+      C = lsb:1;
+      op4_b_dst = op4_c_src >> 1;
+      update_flags_zn(op4_b_dst);
+    }
+    :ror op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x03 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = (op4_c_src >> 1) | (op4_c_src << 31);
+    }
+    :rrc op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x04 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = (op4_c_src >> 1) | (zext(C) << 31);
+    }
+    :rrc.f op4_b_dst, op4_c_src, 1 is (l_ra_raw=0x04 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      local lsb:4 = op4_c_src & 1:4;
+      op4_b_dst = (op4_c_src >> 1) | (zext(C) << 31);
+      C = lsb:1;
+      update_flags_zn(op4_b_dst);
+    }
+    :sexb op4_b_dst, op4_c_src is (l_ra_raw=0x05 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = sext(op4_c_src:1);
+    }
+    :sexb.f op4_b_dst, op4_c_src is (l_ra_raw=0x05 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = sext(op4_c_src:1);
+      update_flags_zn(op4_b_dst);
+    }
+    :sexw op4_b_dst, op4_c_src is (l_ra_raw=0x06 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = sext(op4_c_src:2);
+    }
+    :sexw.f op4_b_dst, op4_c_src is (l_ra_raw=0x06 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = sext(op4_c_src:2);
+      update_flags_zn(op4_b_dst);
+    }
+    :extb op4_b_dst, op4_c_src is (l_ra_raw=0x07 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = zext(op4_c_src:1);
+    }
+    :extb.f op4_b_dst, op4_c_src is (l_ra_raw=0x07 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = zext(op4_c_src:1);
+      update_flags_zn(op4_b_dst);
+    }
+    :extw op4_b_dst, op4_c_src is (l_ra_raw=0x08 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = zext(op4_c_src:2);
+    }
+    :extw.f op4_b_dst, op4_c_src is (l_ra_raw=0x08 & l_flag=1 & op4_b_dst) ... & op4_c_src {
+      op4_b_dst = zext(op4_c_src:2);
+      update_flags_zn(op4_b_dst);
+    }
+    :abs op4_b_dst, op4_c_src is (l_ra_raw=0x09 & l_flag=0 & op4_b_dst) ... & op4_c_src {
+      if (op4_c_src s< 0) goto <negative>;
```

Comment:
Hello, thanks for your comments! I am taking a look at them.
For this specific one, I have a question:
> `op4_b_dst = (sext(op4_c_src s< 0)*op4_c_src & 0x7fffffff) | (sext(op4_c_src s>=0)*op4_c_src);`

This pattern looks nice but I am wondering: why is `sext` used instead of `zext`? Is there some magic that doing `sext` on a boolean expands to `-1` instead of `1`?

Otherwise, this pattern seems wrong for ARCompact (I believe `(-op4_c_src)` should be used instead of `op4_c_src & 0x7fffffff`) but I will properly test my modifications before updating this PR.

Reply:
Done, in https://github.com/NationalSecurityAgency/ghidra/pull/3006/commits/f933b2c8c7ae30844c56a370a0d2157380b6a7f4 : I copied what Tricore did in https://github.com/NationalSecurityAgency/ghidra/blob/b55d1049d4d9bcf66fbb2273b4d46ec95c790834/Ghidra/Processors/tricore/data/languages/tricore.sinc#L615-L618
because this gives more readable decompilation output than `op4_b_dst = (sext(op4_c_src s< 0)*op4_c_src & 0x7fffffff) | (sext(op4_c_src s>=0)*op4_c_src);`